reader-privacy.md

```diff
@@ -0,0 +1,123 @@
+that individually for each CID. A group of CIDs is represented by a unique ContextID as can be seen
+on the diagram below:
+​
+![Index building flow](resources/readers-privacy-1.png)
+
+## Specification
+​
+This specification focuses on improving the **step #3** where a client has to pass a CID to the indexer *in open* 
+to get a list of providers where the content can be fetched from. 
+
+In order to protect the reader's privacy the proposal is to change the way how CID lookup works to the following:
+
+* A client who wants to do a lookup will calculate a hash over the CID (`hash(CID)`) and use it for the
+lookup query (hence the name double hashing);
+* In response to the hashed find request, the indexer will return a set of encrypted `ProviderRecordKey`s. 
+`ProviderRecordKey` will consist of two concatenated hashes - one over `peerID` and the other over `contextID`. 
+Each `ProviderRecordKey` will be encrypted with a key derived from the *original* CID value: 
+`enc(hash(peerID) || hash(contextID), CID)`, where `hash` is a hash over the value, `||` is concatenation 
+and `enc` is encryption over the value. In order to make sense of that payload, a passive observer would need 
+to get hold of the original CID that isn't revealed during the communication round;
+* Using the original CID, the client would decrypt `ProviderRecordKey`s and then calculate another hash
+over the decrypted `hash(peerID)` part of it. Using that hash for each `ProviderRecordKey` the client would do another lookup 
+to get an encrypted `ProviderRecord` in response. `ProviderRecord` will contain information about provider, 
+such as it's *peerID*, *multiaddresses*, *supported protocols* and etc. Each `ProviderRecord` will be encrypted 
+with a key derived from `hash(peerID)`. In order to make sense of that payload, a passive observer would need to 
+get hold of the decrypted `ProviderRecordKey` that isn't revealed during the communication round;
+* Using the `hash(peerID)` from `ProviderRecordKey`s, the client would decrypt `ProviderRecord`s and then reach out to the 
+provider directly to fetch the desired content. 
+
+By utilising such scheme only a party that knows original CID that is being looked up can decode the protocol,
+however that CID is never revealed. 
+
+### Security
+​
+Security model of the Reader's Privacy proposal boils down to inability to *algorithmically* derive the original CID value for a 
+`hash(CID)` that is used for IPNI lookups. Right now indexer advertisments are not encrypted, but authenticated and contain plain CID values in them. 
+That is going to change once *Writer's Privacy* is implemented. Before that a sophisticated attacker could build a map of `hash(CID) -> CID` 
+by re-ingesting advertisements chain from each publisher so that they can use it to decrypt the protocol. 
+Doing that will require significant investment into infrastructure and will be eliminated as a possibility after *Writer's Privacy* upgrade.
```

Comment:
```suggestion
Doing that will require significant resources as it involves crawling the entire network. However, it will eventually be eliminated by *Writer Privacy* upgrade.
```